packages/backend/src/nest/storage/UserProfileStore.ts

```diff
@@ -0,0 +1,261 @@
+      if (!isPng(photoBytes)) {
+        logger.error('Expected PNG for user profile photo', userProfile.pubKey)
+        return false
+      }
+
+      // 200 KB = 204800 B limit
+      //
+      // TODO: Perhaps the compression matters and we should check
+      // actual dimensions in pixels?
+      if (photoBytes.length > 204800) {
+        logger.error('User profile photo must be less than or equal to 200KB')
+        return false
+      }
+    } catch (err) {
+      logger.error('Failed to validate user profile:', userProfile.pubKey, err?.message)
+      return false
+    }
+
+    return true
+  }
+
+  public static async validateUserProfileEntry(entry: LogEntry<UserProfile>) {
+    let verify = false
+
+    try {
+      if (entry.payload.key !== entry.payload.value.pubKey) {
+        logger.error('Failed to verify user profile entry:', entry.hash, 'entry key != payload pubKey')
+        return false
+      }
+      verify = await UserProfileStore.validateUserProfile(entry.payload.value)
+    } catch (err) {
+      logger.error('Failed to verify user profile entry:', entry.hash, err?.message)
+    }
+
+    return verify
+  }
+
+  public async getUserProfiles(): Promise<UserProfile[]> {
+    return Object.values(this.store.all)
+  }
+}
+
+/**
+ * Modified from:
+ * https://github.com/orbitdb/orbit-db-kvstore/blob/main/src/KeyValueIndex.js
+ *
+ * Adds validation function that validates each entry before adding it
+ * to the index.
+ *
+ * TODO: Save latest entry and only iterate over new entries in updateIndex
+ */
+export class KeyValueIndex<T> {
+  private _index: Record<string, any>
+  private validateFn: (entry: LogEntry<T>) => Promise<boolean>
+
+  constructor(validateFn: (entry: LogEntry<T>) => Promise<boolean>) {
+    this._index = {}
+    this.validateFn = validateFn
+  }
+
+  get(key: string) {
+    return this._index[key]
+  }
+
+  async updateIndex(oplog: { values: LogEntry<T>[] }) {
+    const values: LogEntry<T>[] = []
+    const handled: Record<string, boolean> = {}
+
+    for (const v of oplog.values) {
+      if (await this.validateFn(v)) {
+        values.push(v)
+      }
+    }
+
+    for (let i = values.length - 1; i >= 0; i--) {
+      const item = values[i]
+      if (typeof item.payload.key === 'string') {
+        if (handled[item.payload.key]) {
+          continue
+        }
+        handled[item.payload.key] = true
+        if (item.payload.op === 'PUT') {
+          this._index[item.payload.key] = item.payload.value
+          continue
+        }
+        if (item.payload.op === 'DEL') {
+          delete this._index[item.payload.key]
+          continue
+        }
+      } else {
+        logger.error(`Failed to update key/value index - key is not string: ${item.payload.key}`)
+      }
+    }
+  }
+}
+
+export class UserProfileKeyValueIndex extends KeyValueIndex<UserProfile> {
+  constructor() {
+    super(UserProfileStore.validateUserProfileEntry)
+  }
+}
```

Comment:
The bulk of the backend changes are here. I've used a custom index here so that we can filter entries before they are indexed, otherwise invalid entries could overwrite valid entries because the index is a hash map. I'm doing this instead of using an access controller because it looks like access controllers only validate heads and we really need to validate each entry.

Reply:
> Got it, so this is what we'll have to do pretty much whenever we want to validate data?

What do you mean?

> Did we create an issue for orbitdb about this yet?

Not yet, I think Haad would prefer to discuss things like this over Matrix, so I intend to do that soon.

Reply:
>> Got it, so this is what we'll have to do pretty much whenever we want to validate data?
> What do you mean?

I mean, it looks like we're iterating through every item and validating it, is that right? When do we have to do our own validation in this way? On messages? On channels? 

> Not yet, I think Haad would prefer to discuss things like this over Matrix, so I intend to do that soon.

Yes let's ask soon!

Reply:
> I mean, it looks like we're iterating through every item and validating it, is that right? When do we have to do our own validation in this way? On messages? On channels?

Yes that's what we are doing. I think if you want to make sure you validate everything, then this is one way to do it.